### PR TITLE
fix(phase-01): negative cross-entropy, all-zero stationary distribution, inverted Cohen's d

### LINKED ISSUE #343: [bug] Phase 1: negative cross-entropy, all-zero stationary distribution, and an inverted Cohen's d
## Where

- Phase / lesson: Phase 1 · 13-numerical-stability, 22-stochastic-processes, 15-statistics-for-ml
- Files: `code/numerical.py`, `code/stochastic.py`, `code/statistics.py`

## What's wrong

Three wrong-math defects in Phase 1. Two of them print visibly wrong numbers; the third contradicts the lesson's own docs.

**1. `13-numerical-stability` — `binary_cross_entropy_stable` returns *negative* cross-entropy.**
The function takes a logit, so it should compute `softplus(x) - y*x`. With `max_val = max(0, -logit)`, the first term is the stable form of `log(1+e^-x)`, i.e. `softplus(-x)`, not `softplus(x)`. Since `softplus(-x) = softplus(x) - x`, every result is short by exactly the logit:

| logit | y | `naive(y, sigmoid(x))` | `stable(y, x)` |
|---|---|---|---|
| 0.5 | 1 | 0.474077 | **-0.025923** |
| 2.0 | 1 | 0.126928 | **-1.873072** |
| 6.0 | 1 | 0.002476 | **-5.997524** |

A cross-entropy of -5.9975 is impossible. For `y=0` it silently returns the other label's loss. No demo calls the function, which is why this went unnoticed.

**2. `22-stochastic-processes` — the stationary distribution prints `0.0000` for every state.**
`stationary_distribution` clips the lambda=1 eigenvector at 0. LAPACK's eigenvector sign is arbitrary, and for this lesson's weather matrix it comes back all-negative, so clipping zeroes every entry, the sum is 0, the `if total > 0` guard skips normalisation, and the all-zero vector is returned:

```
Stationary distribution (analytical):
  Sunny: 0.0000   Rainy: 0.0000   Cloudy: 0.0000
Empirical distribution (100000 steps, start=Sunny):
  Sunny: 0.5469   Rainy: 0.1826   Cloudy: 0.2705
```

`docs/en.md:94` promises roughly `[0.53, 0.18, 0.29]`, and the convergence check below it never converges (max error stays ~0.55). Perron-Frobenius guarantees the vector is single-signed, so `abs` is the correct normalisation — which is what `docs/en.md` itself shows.

**3. `15-statistics-for-ml` — `cohens_d` sign is inverted.**
`docs/en.md:301` defines `Cohen's d = (mean_1 - mean_2) / pooled_std` and the sibling `t_statistic_two_sample` returns `(m1 - m2) / se`, but `cohens_d` returns `(m2 - m1) / pooled`. The demo prints both for the same pair on adjacent lines, so the output reads `t-statistic: -2.4599` then `Cohen's d: 0.6351`. Magnitude is right and `interpret_cohens_d` uses `abs()`, so no verdict changes — but the file contradicts itself and the docs.

## Reproduce

1. `python3 -c "import numerical; print(numerical.binary_cross_entropy_stable(1.0, 6.0))"` → `-5.9975`, should be `0.0025`.
2. `python3 phases/01-math-foundations/22-stochastic-processes/code/stochastic.py` → DEMO 2 prints all zeros.
3. `python3 phases/01-math-foundations/15-statistics-for-ml/code/statistics.py` → t-statistic and Cohen's d print with opposite signs.

## Environment

- OS: macOS 15
- Python 3.9.6, numpy 2.0.2
- How you ran it: local

I have one-line fixes for all three and will open a PR referencing this issue.

### PR DESCRIPTION
**Pull request by thejesh23**
## What this PR does

Three one-line fixes for wrong-math defects in Phase 1. Fixes #343.

## Kind of change

- [x] Fix to an existing lesson

## Checklist

- [x] Code runs without errors with the listed dependencies
- [x] No comments in code files (docs explain, code is self-explanatory)
- [x] One lesson per commit (atomic per-lesson rule)
- [x] Tested locally / code output matches what `docs/en.md` claims

## Phase / lesson

Phase 1 · 13-numerical-stability, 22-stochastic-processes, 15-statistics-for-ml

## The three fixes

**`13-numerical-stability` — `binary_cross_entropy_stable` returned negative cross-entropy.**
The function takes a logit, so it should compute `softplus(x) - y*x`. With `max_val = max(0, -logit)` the first term is the stable form of `log(1+e^-x)` = `softplus(-x)`, so every result was short by exactly the logit — `BCE(y=1, logit=6)` came out as `-5.9975` instead of `0.0025`. Pairing that term with `+(1-y)*x` gives the correct value and matches the standard BCEWithLogits form.

Round-trip check after the fix (not just "the negative numbers are gone"):

```
stable(y, x) == log1p(exp(-|x|)) + max(x,0) - y*x   for x in {-800,-10,-3,0,0.5,2,6,800}, y in {0,1}: True
stable(y, x) == binary_cross_entropy_naive(y, sigmoid_stable(x)):                                    True
```

It also stays finite at +-800, where the naive form overflows — which is the point of the lesson.

**`22-stochastic-processes` — stationary distribution printed `0.0000` for every state.**
LAPACK returns the lambda=1 eigenvector with an arbitrary sign; for this weather matrix it is all-negative, so `np.clip(v, 0, None)` zeroed it, the sum became 0, and the `if total > 0` guard skipped normalisation. Perron-Frobenius guarantees the vector is single-signed, so `np.abs` is the correct normalisation — and it is what `docs/en.md` already shows.

Before: `Sunny 0.0000 / Rainy 0.0000 / Cloudy 0.0000` against an empirical `0.5469 / 0.1826 / 0.2705`.
After: `Sunny 0.5455 / Rainy 0.1818 / Cloudy 0.2727` — exactly 6/11, 2/11, 3/11, matching `docs/en.md:94`. The convergence check below it now actually converges.

**`15-statistics-for-ml` — `cohens_d` sign matched neither the docs nor its sibling.**
`docs/en.md:301` defines `d = (mean_1 - mean_2) / pooled_std` and `t_statistic_two_sample` returns `(m1 - m2) / se`, but `cohens_d` returned `(m2 - m1) / pooled`, so the demo printed the two with opposite signs on adjacent lines. After: `cohens_d([1,2,3,4,5],[3,4,5,6,7])` = `-1.2649`, agreeing with `t` = `-2.0`.

## Notes for reviewer

Fix 3 changes only the reported sign — magnitudes are unchanged and `interpret_cohens_d` already takes `abs()`, so no verdict in the demo output flips.

While auditing the same file I also hit two quadrature problems in `statistics.py` that are **not** in this PR because the fix is a real rewrite rather than a one-liner: `_lower_incomplete_gamma_ratio` returns `p = 0.0352` for chi2=8, df=1 where `docs/en.md:247` states `p < 0.005` (exact: 0.00468), and `_regularized_beta` collapses to `p = 0.000000` at large df where the true value is 0.0636 — which is what makes the "statistical vs practical significance" demo print `Significant: True`. Happy to open a separate issue/PR for those with a continued-fraction implementation if you want them fixed.

### CODE REVIEW

**Comment by thejesh23** [comment]
Superseded by d1cb9d1 — closing.

Verified all three on `main`:

- **`13-numerical-stability`** — `max_val = max(0.0, logit)` with `exp(logit - max_val)` computes `softplus(x)` stably, so the result is `softplus(x) - y*x`. Different arrangement from my patch, same correct value, and it keeps the max-subtraction the lesson is about.
- **`22-stochastic-processes`** — sign-flip on `stationary.sum() < 0` before the clip. Handles the all-negative eigenvector case that produced the all-zero output.
- **`15-statistics-for-ml`** — `(m1 - m2) / pooled`, identical to this patch.

Thanks.
